\chapter{Examples of Biological Information Files}\label{annex:formats}

\section{SAM Example}\label{annex:sam}
\begin{lstlisting}[numbers=none, breaklines=true]
  HD	VN:1.0 SO:coordinate
  @SQ	SN:seq1	LN:5000
  @SQ	SN:seq2	LN:5000
  @CO	Example of SAM/BAM file format.
  B7_591:4:96:693:509	73	seq1	1	99	36M	*	0	0	CACTAGTGGCTCATTGTAAATGTGTGGTTTAACTCG	<<<<<<<<<<<<<<<;<<<<<<<<<5<<<<<;:<;7	MF:i:18	Aq:i:73	NM:i:0	UQ:i:0	H0:i:1	H1:i:0
  EAS54_65:7:152:368:113	73	seq1	3	99	35M	*	0	0	CTAGTGGCTCATTGTAAATGTGTGGTTTAACTCGT	<<<<<<<<<<0<<<<655<<7<<<:9<<3/:<6):	MF:i:18	Aq:i:66	NM:i:0	UQ:i:0	H0:i:1	H1:i:0
  EAS51_64:8:5:734:57	137	seq1	5	99	35M	*	0	0	AGTGGCTCATTGTAAATGTGTGGTTTAACTCGTCC	<<<<<<<<<<<7;71<<;<;;<7;<<3;);3*8/5	MF:i:18	Aq:i:66	NM:i:0	UQ:i:0	H0:i:1	H1:i:0
  B7_591:1:289:587:906	137	seq1	6	63	36M	*	0	0	GTGGCTCATTGTAATTTTTTGTTTTAACTCTTCTCT	(-&----,----)-)-),'--)---',+-,),''*,	MF:i:130	Aq:i:63	NM:i:5	UQ:i:38	H0:i:0	H1:i:0
  EAS56_59:8:38:671:758	137	seq1	9	99	35M	*	0	0	GCTCATTGTAAATGTGTGGTTTAACTCGTCCATGG	<<<<<<<<<<<<<<<;<;7<<<<<<<<7<<;:<5%	MF:i:18	Aq:i:72	NM:i:0	UQ:i:0	H0:i:1	H1:i:0
  EAS56_61:6:18:467:281	73	seq1	13	99	35M	*	0	0	ATTGTAAATGTGTGGTTTAACTCGTCCCTGGCCCA	<<<<<<<<;<<<8<<<<<;8:;6/686&;(16666	MF:i:18	Aq:i:39	NM:i:1	UQ:i:5	H0:i:0	H1:i:1
  EAS114_28:5:296:340:699	137	seq1	13	99	36M	*	0	0	ATTGTAAATGTGTGGTTTAACTCGTCCATGGCCCAG	<<<<<;<<<;<;<<<<<<<<<<<8<8<3<8;<;<0;	MF:i:18	Aq:i:73	NM:i:0	UQ:i:0	H0:i:1	H1:i:0
\end{lstlisting}

\section{VCF Example}\label{annex:vcf}
\begin{lstlisting}[numbers=none, breaklines=true]
  ##fileformat=VCFv4.0
  ##fileDate=20090805
  ##source=myImputationProgramV3.1
  ##reference=1000GenomesPilot-NCBI36
  ##phasing=partial
  ##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data">
  ##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
  ##INFO=<ID=AF,Number=.,Type=Float,Description="Allele Frequency">
  ##INFO=<ID=AA,Number=1,Type=String,Description="Ancestral Allele">
  ##INFO=<ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129">
  ##INFO=<ID=H2,Number=0,Type=Flag,Description="HapMap2 membership">
  ##FILTER=<ID=q10,Description="Quality below 10">
  ##FILTER=<ID=s50,Description="Less than 50% of samples have data">
  ##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
  ##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
  ##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
  ##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality">
  #CHROM POS     ID        REF ALT    QUAL FILTER INFO                              FORMAT      NA00001        NA00002        NA00003
  20     14370   rs6054257 G      A       29   PASS   NS=3;DP=14;AF=0.5;DB;H2           GT:GQ:DP:HQ 0|0:48:1:51,51 1|0:48:8:51,51 1/1:43:5:.,.
  20     17330   .         T      A       3    q10    NS=3;DP=11;AF=0.017               GT:GQ:DP:HQ 0|0:49:3:58,50 0|1:3:5:65,3   0/0:41:3
  20     1110696 rs6040355 A      G,T     67   PASS   NS=2;DP=10;AF=0.333,0.667;AA=T;DB GT:GQ:DP:HQ 1|2:21:6:23,27 2|1:2:0:18,2   2/2:35:4
  20     1230237 .         T      .       47   PASS   NS=3;DP=13;AA=T                   GT:GQ:DP:HQ 0|0:54:7:56,60 0|0:48:4:51,51 0/0:61:2
  20     1234567 microsat1 GTCT   G,GTACT 50   PASS   NS=3;DP=9;AA=G                    GT:GQ:DP    0/1:35:4 
\end{lstlisting}

\section{FASTQ Example}\label{annex:fastq}
\begin{lstlisting}[numbers=none, breaklines=true]
  @SRR014849.1 EIXKN4201CFU84 length=93
  GGGGGGGGGGGGGGGGCTTTTTTTGTTTGGAACCGAAAGG
  GTTTTGAATTTCAAACCCTTTTCGGTTTCCAACCTTCCAA
  AGCAATGCCAATA
  +SRR014849.1 EIXKN4201CFU84 length=93
  3+&$#"""""""""""7F@71,'";C?,B;?6B;:EA1EA
  1EA5'9B:?:#9EA0D@2EA5':>5?:%A;A8A;?9B;D@
  /=<?7=9<2A8==
\end{lstlisting}

\section{FASTA Example}\label{annex:fasta}
\begin{lstlisting}[numbers=none, breaklines=true]
  >gi|5524211|gb|AAD44166.1| cytochrome b [Elephas maximus maximus]
  LCLYTHIGRNIYYGSYLYSETWNTGIMLLLITMATAFMGYVLPWGQMSFWGATVITNLFSAIPYIGTNLV
  EWIWGGFSVDKATLNRFFAFHFILPFTMVALAGVHLTFLHETGSNNPLGLTSDSDKIPFHPYYTIKDFLG
  LLILILLLLLLALLSPDMLGDPDNHMPADPLNTPLHIKPEWYFLFAYAILRSVPNKLGGVLALFLSIVIL
  GLMPFLHTSKHRSMMLRPLSQALFWTLTMDLLTLTWIGSQPVEYPYTIIGQMASILYFSIILAFLPIAGX
  IENY
\end{lstlisting}

\section{GTF/GFF Example}\label{annex:gff}
\begin{lstlisting}[numbers=none, breaklines=true]
  ## GFF
  ##gff-version 3
  ##sequence-region   ctg123 1 1497228
  ctg123 . gene            1000  9000  .  +  .  ID=gene00001;Name=EDEN
  ctg123 . TF_binding_site 1000  1012  .  +  .  ID=tfbs00001;Parent=gene00001
  ctg123 . mRNA            1050  9000  .  +  .  ID=mRNA00001;Parent=gene00001;Name=EDEN.1
  ctg123 . mRNA            1050  9000  .  +  .  ID=mRNA00002;Parent=gene00001;Name=EDEN.2
  ctg123 . mRNA            1300  9000  .  +  .  ID=mRNA00003;Parent=gene00001;Name=EDEN.3
  ctg123 . exon            1300  1500  .  +  .  ID=exon00001;Parent=mRNA00003
  ctg123 . exon            1050  1500  .  +  .  ID=exon00002;Parent=mRNA00001,mRNA00002
  ctg123 . exon            3000  3902  .  +  .  ID=exon00003;Parent=mRNA00001,mRNA00003

  ---------------------------------------------------

  ## GTF
  140    Twinscan    inter         5141     8522     .    -    .    gene_id ""; transcript_id "";
  140    Twinscan    inter_CNS     8523     9711     .    -    .    gene_id ""; transcript_id "";
  140    Twinscan    inter         9712     13182    .    -    .    gene_id ""; transcript_id "";
  140    Twinscan    3UTR          65149    65487    .    -    .    gene_id "140.000"; transcript_id "140.000.1";
  140    Twinscan    3UTR          66823    66992    .    -    .    gene_id "140.000"; transcript_id "140.000.1";
  140    Twinscan    stop_codon    66993    66995    .    -    0    gene_id "140.000"; transcript_id "140.000.1";
\end{lstlisting}
