\chapter{Introduction} \label{chap:intro}

\section*{}

%\begin{Notes}
%- Introduction seems fine, check later if emphasis on sequencing should be toned
%down.
%\end{Notes}

Molecular biology is a branch of biology that studies biological activities of
living beings, at a molecular level. The grounds for this field of study were
set in the early 1930s, although it only emerged in its modern form in the
1960s, with the discovery of the structure of \dna. Among the processes studied
by this branch of biology is gene expression. Gene expression (further explained
in Chapter \ref{chap:sota}) is the process by which \dna{} molecules are
transformed into useful genetic products, typically proteins, which are
essential for living organisms. This knowledge is not only important in fields
like evolutionary or molecular biology, but has crucial applications in fields
such as medicine. One example of such an application is the usage of gene
expression analysis in the diagnosis and treatment of cancer patients
\cite{Pusztai01062003}.

With the advent of \ngs{} (\textit{Next Generation Sequencing}) techniques,
researchers have at their disposal huge amounts of sequencing data, that is not
only cheaper and faster to produce, but also more commonly available. This data
can then be used to obtain relevant information about organisms' gene
expression. But, as the cost of sequencing genomes was reduced, the cost of
processing such information was increased. \ngs{} techniques tend to produce
much smaller reads\footnote{A \textit{read} is a single fragment of a
genome/\trans, obtained through sequencing techniques.} than previously used
techniques, presenting a more challenging problem, from a computational
standpoint \cite{Wolf2013}.

\section{Domain Problem} \label{sec:problem}

\begin{Notes}
- Check references to sections.\\
\end{Notes}

Despite its great advancements in the past decades, molecular biology is still a
relatively new subject and, as such, there are still some unknowns and partial
knowledge in this area. In respect to gene expression, some mechanisms of this
intricate process are yet to be fully understood. One such mechanism is the one
that regulates the transcription speed into \rna. One of the objectives of this
thesis is to understand how the final sequences of a gene's exons are
responsible for the speed at which the exons themselves are transcribed. The
other objective is to understand how RNA-binding protein (RBP) manipulation can
be used to better understand an organism's gene expression. These are, however,
complex tasks that can be further decomposed in the three main problems that
will be addressed in the thesis, namely:

\begin{itemize}

  \item
  Sequencing read alignment against a reference genome and differential
  expression analysis between samples of different individuals (of the same
  species). This is effectively one of the most complex problems addressed in
  the thesis. We will use data obtained through a sequencing method called
  \rnaseq{}\footnote{\rnaseq{} is also referred to as \textit{Whole \Trans{}
  Shotgun Sequencing}, or WTSS.}. Further insight about this method will be
  given in Chapter \ref{chap:sota}, with particular emphasis for tools used to
  align and analyze this data (Section \ref{sec:seqtools}).

  \item
  Gene enrichment and RBP analysis. This part of the work aims to collect as
  much relevant information as possible about the particular genes being studied
  at the time, to help biologists to better understand their function. RBP
  knowledge is particularly important for gene manipulation and a very useful
  tool for better understanding gene expression, as will be further described in
  Chapter \ref{chap:sota}.

  \item
  Further analysis of the produced data, using machine learning techniques for
  data mining, specifically for clustering analysis. These techniques will be
  employed in an effort to give biologists more relevant information about gene
  expression, uncovering possible relationships in the retrieved information.
  This topic will be developed in Section \ref{sec:mlearning}.

\end{itemize}

Solving these problems requires the use of computational tools. As such, the
development of a computer system (or multiple systems) to tackle these problems
emerges as a secondary objective of the thesis. The details of the design of
this system will be presented in Chapter \ref{chap:description}, while its
concrete implementation will be discussed in Chapter \ref{chap:implementation}.

\section{Motivation and Objectives} \label{sec:motivation}

%\begin{Notes}
%- Check previous notes, genome assembly is not needed.\\
%- Don't forget to mention the gene enrichment/protein binding site analysis.
%- Talk about how simplicity is needed because many times biologists don't
%understand the tools already available.
%\end{Notes}

Gene expression analysis is essential for modern day molecular biology. Among
many of the possible applications of this information, we can highlight: better
classification and diagnosis of diseases, assessing how cells react to a
specific treatment, and others.

While nowadays powerful computational tools exist to target almost any biology
problem, many of those tools require a very specific set of technical skills and
have a steep learning curve. Possibly the most important motivation behind this
thesis, and ultimately its main objective, is to provide researchers with
powerful yet simple and user friendly tools. This means developing a system
simple enough that any user can learn to operate it in a short period of time
with minimal effort, but sufficiently advanced to suit the user's research
needs.

Another typical problem that biology researchers face nowadays is information
dispersion and the repetitive and lengthy task of compiling that information.
Researchers frequently have to manually join information originating from a
multitude of different platforms, which use inconsistent formats and notations.
Our second objective is therefore to provide a system that is able to take this
burden off the user, making the process faster and simpler.

\section{Project} \label{sec:project}

\begin{Notes}
- Add chapter/section references.\\
\end{Notes}

The project itself revolves around the development of a prototype computer
system, capable of solving the aforementioned problems. Due to the complexity of
the complete system, its development followed a modular organization (further
described in Chapter \ref{chap:description}). The envisioned system
architecture is divided into three major components:

\begin{description}

  \item[Differential expression analysis pipeline]
  is responsible for aligning reads against a reference genome and compare
  contrasts between different samples. The pipeline is based on the preexisting
  iRAP pipeline\footnote{\url{https://code.google.com/p/irap/}}. The pipeline's
  capabilities are further enhanced with both job configuration automation and
  differential expression results consolidation (combining results from multiple
  differential expression tools).

  \item[RNA-binding protein analysis workflow]
  aggregates information about RBPs from multiple biologic web databases
  (Ensembl, NCBI, UniProt, etc.) and organizes it in ways that are useful to
  biology researchers. Moreover, this information is clustered using data mining
  techniques, in order to reveal groups of genes and RBPs that may hold biologic
  relevance.

  \item[Web platform]
  is responsible for storing and managing genetic data, coordinating interaction
  between the other components of the system and providing a web interface for
  user interaction. This component is based mainly on typical web technologies,
  that is, a document based database for data storage (MongoDB), a web framework
  for business logic implementation (Padrino) and web markup and styling
  languages for interface implementation (HTML, CSS).

\end{description}

\section{Structure of the Report} \label{sec:outline}

\begin{Notes}
- Update this last.\\
\end{Notes}

Besides the introduction chapter, this document is composed by three additional
chapters. Chapter \ref{chap:sota} introduces some basic biology and RNA-Seq
concepts, that are essential to understand the problems with which this document
deals. Furthermore, we describe the main techniques used for genome/\trans{}
sequencing and assembly, their differences, applications and the tools and data
formats typically used in those areas. Lastly, we give some insight about data
mining algorithms and how they will be applied in the context of the project.
Chapter \ref{chap:workplan} outlines the main steps in the development of the
project (and the respective software prototype) and attempts to provide a
feasible schedule for the work's execution. It also presents the datasets that
will be studied and used in this work, their origins and features, as well as
the validation methods that will be used to ascertain the quality of our
results. Chapter \ref{chap:conclusions} sums up the what has been defined in the
report, emphasizing the problem that the thesis addresses and the work that will
be executed towards solving that problem. It will also give a brief idea of what
are the expected results at the end of the project.
